### PR TITLE
Add swiftmailer listener plugin

### PR DESCRIPTION
**Pull request by Prokyonn**
| Q | A
| --- | ---
| Bug fix? | no
| New feature? | no
| BC breaks? | no
| Deprecations? | no
| Fixed tickets | fixes #issuenum
| Related issues/PRs | #issuenum
| License | MIT
| Documentation PR | sulu/sulu-docs#prnum

#### What's in this PR?

Adds a markup listener for `sulu-link` as a swiftmailer plugin

#### Why?

The body of e-mails sent with the swiftmailer are not parsed by the markup bundle and thus the `sulu-link` tag is not replaced with the correct `<a>` element.

### CODE REVIEW

File: src/Sulu/Bundle/MarkupBundle/Listener/SwiftMailerListener.php [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+        $message = $event->getMessage();
+
+        $body = $message->getBody();
+        $format = $message->getBodyContentType();
+
+        if (count($explodedFormat = explode('/', $format)) > 1) {
+            $format = $explodedFormat[1];
+        }
+
+        $locale = $this->requestStack->getCurrentRequest()->getLocale();
```
Comment (alexander-schranz) [inline src/Sulu/Bundle/MarkupBundle/Listener/SwiftMailerListener.php:48]: `getCurrentRequest` could be empty (e.g. emails send from cli) we need to fallback here to the default (%kernel.locale%).

File: src/Sulu/Bundle/MarkupBundle/DependencyInjection/CompilerPass/ParserCompilerPass.php [hunk 2]
```diff
@@ -20,29 +20,34 @@
  */
 class ParserCompilerPass implements CompilerPassInterface
 {
-    const SERVICE_ID = 'sulu_markup.response_listener';
+    const SERVICE_IDS = [
+        'sulu_markup.response_listener',
+        'sulu_markup.swift_mailer_listener'
```
Comment (alexander-schranz) [inline src/Sulu/Bundle/MarkupBundle/DependencyInjection/CompilerPass/ParserCompilerPass.php:25]: for new services we should use something like here: https://github.com/sulu/SuluHeadlessBundle/pull/20/files#diff-66e8662ac6150c33e219664495191cf6R105

so no changes in this compilerpass should be needed

File: src/Sulu/Bundle/MarkupBundle/Tests/Unit/Listener/SwiftMailerListenerTest.php [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+
+        $this->listener = new SwiftMailerListener(['html' => $this->markupParser->reveal()], $this->requestStack->reveal());
+    }
+
+    public function testReplaceMarkup(): void
+    {
+        /** @var Request|ObjectProphecy $request */
+        $request = $this->prophesize(Request::class);
+        $request->getLocale()->willReturn('de');
+        $this->requestStack->getCurrentRequest()->willReturn($request->reveal());
```
Comment (alexander-schranz) [inline src/Sulu/Bundle/MarkupBundle/Tests/Unit/Listener/SwiftMailerListenerTest.php:70]: add test where request is null

File: src/Sulu/Bundle/MarkupBundle/Markup/HtmlMarkupParser.php [hunk 4]
```diff
@@ -60,12 +60,17 @@ public function validate($content, $locale)
...
+            $tags = $this->tagRegistry->getTag($tagMatchGroup->getTagName(), $this->getType(), $tagMatchGroup->getNamespace())
                 ->validateAll($tagMatchGroup->getTags(), $locale);
 
             $result = array_merge($result, $tags);
         }
 
         return $result;
     }
+
+    public function getType(): string
```
Comment (Prokyonn) [inline src/Sulu/Bundle/MarkupBundle/Markup/HtmlMarkupParser.php:72]: @alexander-schranz I guess we cannot take this method into the interface because of bc right?

Comment (alexander-schranz) [reply src/Sulu/Bundle/MarkupBundle/Markup/HtmlMarkupParser.php:72]: Would not add this method just index by the key of the tagged service. See https://symfony.com/blog/new-in-symfony-4-3-indexed-and-tagged-service-collections

File: src/Sulu/Bundle/MarkupBundle/Tests/Unit/Listener/SwiftMailerListenerTest.php [hunk 5]
```diff
@@ -0,0 +1,107 @@
...
+        $this->simpleMessage->setBody('<html><a href="/test">Page-Title</a></html>')->shouldBeCalled();
+
+        $this->listener->beforeSendPerformed($this->event->reveal());
+    }
+
+    public function testReplaceMarkupWithoutRequest(): void
+    {
+        $this->requestStack->getCurrentRequest()->willReturn(null);
+
+        $this->simpleMessage->getBodyContentType()->wilLReturn('text/html');
```
Comment (alexander-schranz) [inline src/Sulu/Bundle/MarkupBundle/Tests/Unit/Listener/SwiftMailerListenerTest.php:97]: `wilL`

File: src/Sulu/Bundle/MarkupBundle/Tests/Unit/Listener/SwiftMailerListenerTest.php [hunk 6]
```diff
@@ -0,0 +1,107 @@
...
+
+        $this->markupParser->parse('<html><sulu-link href="123-123-123"/></html>', 'de')
+            ->willReturn('<html><a href="/test">Page-Title</a></html>')->shouldBeCalled();
+
+        $this->simpleMessage->setBody('<html><a href="/test">Page-Title</a></html>')->shouldBeCalled();
+
+        $this->listener->beforeSendPerformed($this->event->reveal());
+    }
+
+    public function testReplaceMarkupWithoutRequest(): void
```
Comment (alexander-schranz) [inline src/Sulu/Bundle/MarkupBundle/Tests/Unit/Listener/SwiftMailerListenerTest.php:93]: we should also add a test with a unkown `contentType` e.g.: `text/plain`

File: src/Sulu/Bundle/MarkupBundle/Tests/Unit/Listener/SwiftMailerListenerTest.php [hunk 7]
```diff
@@ -0,0 +1,107 @@
...
+    }
+
+    public function testReplaceMarkup(): void
+    {
+        /** @var Request|ObjectProphecy $request */
+        $request = $this->prophesize(Request::class);
+        $request->getLocale()->willReturn('de');
+        $this->requestStack->getCurrentRequest()->willReturn($request->reveal());
+
+        $this->simpleMessage->getBodyContentType()->wilLReturn('text/html');
```
Comment (alexander-schranz) [inline src/Sulu/Bundle/MarkupBundle/Tests/Unit/Listener/SwiftMailerListenerTest.php:82]: `wilL`

File: src/Sulu/Bundle/MarkupBundle/Tests/Unit/Listener/SwiftMailerListenerTest.php [hunk 8]
```diff
@@ -0,0 +1,125 @@
...
+        $request->getLocale()->willReturn('de');
+        $this->requestStack->getCurrentRequest()->willReturn($request->reveal());
+
+        $this->simpleMessage->getBodyContentType()->willReturn('text/plain');
+        $this->simpleMessage->getBody()->willReturn('<html><sulu-link href="123-123-123"/></html>');
+
+        $this->markupParser->parse('<html><sulu-link href="123-123-123"/></html>', 'de')
+            ->shouldNotBeCalled();
+
+        $this->simpleMessage->setBody('<html><a href="/test">Page-Title</a></html>')->shouldNotBeCalled();
```
Comment (alexander-schranz) [inline src/Sulu/Bundle/MarkupBundle/Tests/Unit/Listener/SwiftMailerListenerTest.php:121]: why does it still get replaced its unkown response type?

Comment (alexander-schranz) [reply src/Sulu/Bundle/MarkupBundle/Tests/Unit/Listener/SwiftMailerListenerTest.php:121]: Oh sorry did not see shouldNotBeCalled 🙈

**Comment by alexander-schranz** [comment]
Code looks good for me need to test it in combination with form bundle.

**Comment by alexander-schranz** [comment]
Works, the only sad thing is that the symfony profiler does not show the replaced link but think we can do there nothing.
